src/Database/LMDB/Simple/Cursor.hs

```diff
@@ -0,0 +1,211 @@
+import           Data.Map.Strict               (Map)
+import qualified Data.Map.Strict               as Map
+import           Foreign                       (Ptr, alloca)
+
+import           Database.LMDB.Raw
+import           Database.LMDB.Simple.Codec
+import           Database.LMDB.Simple.Internal hiding (get, peekVal, put)
+
+{-------------------------------------------------------------------------------
+  The Cursor monad
+-------------------------------------------------------------------------------}
+
+-- | An environment used in the cursor monad @'CursorM'@.
+data CursorEnv k v = CursorEnv {
+    -- | A pointer to a lower-level cursor provided by the LMDB Haskell binding.
+    cursor :: MDB_cursor'
+    -- | A pointer that can hold a key.
+  , kPtr   :: Ptr MDB_val
+    -- | A pointer that can hold a value.
+  , vPtr   :: Ptr MDB_val
+    -- | A codec for encoding/decoding keys in CBOR format.
+  , kCodec :: Codec k
+    -- | A codec for encoding/decoding values in CBOR format.
+  , vCodec :: Codec v
+  }
+
+-- | The cursor monad is a @'ReaderT'@ transformer on top of @'IO'@.
+--
+-- We perform lower-level cursor operations in @'IO'@, while the @'ReaderT'@
+-- passes along a @'CursorEnv'@ that contains all the information we need to
+-- perform the lower-level cursor operations.
+newtype CursorM k v a = CursorM {unCursorM :: ReaderT (CursorEnv k v) IO a}
+  deriving stock (Functor)
+  deriving newtype (Applicative, Monad)
+
+-- | Run a cursor monad in @'IO'@.
+runCursorM :: CursorM k v a -> CursorEnv k v -> IO a
+runCursorM cm = runReaderT (unCursorM cm)
+
+-- | Run a cursor monad as a @'Transaction'.
+runCursorAsTransaction ::
+     Codec k            -- ^ A @'Codec'@ for the key type.
+  -> Codec v            -- ^ A @'Codec'@ for the value type.
+  -> CursorM k v a      -- ^ The cursor monad to run.
+  -> Database k v       -- ^ The database to run the cursor monad in.
+  -> Transaction mode a
+runCursorAsTransaction kcod vcod cm (Db _ dbi) = Txn $ \txn ->
+  alloca $ \kptr ->
+    alloca $ \vptr ->
+      withCursor txn dbi (\c -> runCursorM cm (CursorEnv c kptr vptr kcod vcod))
+
+{-------------------------------------------------------------------------------
+  Monadic operations
+-------------------------------------------------------------------------------}
+
+-- | Retrieve a key-value pair using a cursor Get operation.
+--
+-- The different typs of cursor Get operations are listed here:
+-- http://www.lmdb.tech/doc/group__mdb.html#ga1206b2af8b95e7f6b0ef6b28708c9127
+cget :: MDB_cursor_op -> CursorM k v (Maybe (k, v))
+cget op = CursorM $ do
+  r <- ask
+  found <- lift $ mdb_cursor_get' op (cursor r) (kPtr r) (vPtr r)
+  if found then do
+    k <- lift $ peekVal (kCodec r) (kPtr r)
+    v <- lift $ peekVal (vCodec r) (vPtr r)
+    pure $ Just (k, v)
+  else
+    pure Nothing
+
+-- | Like @'cget'@, but throws away the result.
+cget_ :: MDB_cursor_op -> CursorM k v ()
+cget_ = void . cget
+
+cpokeKey :: k -> CursorM k v ()
```

Comment:
I wonder why `pokeKey` is implemented in terms of `pokeVal` :thinking:

Reply:
In the lower-level bindings, keys and values are both `MDB_val`s, which is somewhat confusing I agree -- I will do some renaming to